niko/go-sdk-docs (#553)
Eppo-exp/eppo-docs: docs/sdks/server-sdks/go/bandits.mdx

  name="BanditKey"
  type="string"
  defaultValue='""'
>

The key (unique identifier) of the bandit. Example: `"ad-bandit-1"`
</ApiOptionRef>

<ApiOptionRef 
  name="Subject"
  type="string"
  defaultValue='""'
>

An identifier of the subject or user assigned to the experiment variation. Example: `"ed6f85019080"`
</ApiOptionRef>

<ApiOptionRef 
  name="SubjectNumericAttributes"
  type="map[string]float64"
  defaultValue="nil"
>

Metadata about numeric attributes of the subject. Example: `{"age": 30}`
</ApiOptionRef>

<ApiOptionRef 
  name="SubjectCategoricalAttributes"
  type="map[string]string"
  defaultValue="nil"
>

Metadata about non-numeric attributes of the subject. Example: `{"loyalty_tier": "gold"}`
</ApiOptionRef>

<ApiOptionRef 
  name="Action"
  type="string"
  defaultValue='""'
>

The action assigned by the bandit. Example: `"promo-20%-off"`
</ApiOptionRef>

<ApiOptionRef 
  name="ActionNumericAttributes"
  type="map[string]float64"
  defaultValue="nil"
>

Metadata about numeric attributes of the assigned action. Example: `{"discount": 0.2}`
</ApiOptionRef>

<ApiOptionRef 
  name="ActionCategoricalAttributes"
  type="map[string]string"
  defaultValue="nil"
>

> Metadata about non-numeric attributes of the assigned action. Example: `{"promoTextColor": "white"}`

Reviewer: ```suggestion
Metadata about categorical attributes of the assigned action. Example: `{"promoTextColor": "white"}`
```

</ApiOptionRef>

<ApiOptionRef 
  name="ActionProbability"
  type="float64"
  defaultValue="0"
>

The weight between 0 and 1 the bandit valued the assigned action. Example: `0.25`
</ApiOptionRef>

<ApiOptionRef 
  name="ModelVersion"
  type="string"
  defaultValue='""'
>

Unique identifier for the version of the bandit parameters used. Example: `"v123"`
</ApiOptionRef>

> 

Reviewer: missing:
- OptimalityGap
- MetaData

Author: Adding.

## Querying the Bandit

To query the bandit for an action, use the `GetBanditAction` method:

```go

> attributes := eppoclient.Attributes{

Reviewer: ```suggestion
attributes := eppoclient.ContextAttributes{
```

    NumericAttributes: map[string]float64{
        "age": float64(user.Age),
    },
    CategoricalAttributes: map[string]string{
        "country": user.Country,
    },
}

> actions := map[string]eppoclient.Attributes{

Reviewer: ```suggestion
actions := map[string]eppoclient.ContextAttributes{
```

    "nike": {
        NumericAttributes: map[string]float64{
            "brand_affinity": 2.3,
        },
        CategoricalAttributes: map[string]string{
            "previously_purchased": "true",
        },
    },
    "adidas": {
        NumericAttributes: map[string]float64{
            "brand_affinity": 0.2,
        },
        CategoricalAttributes: map[string]string{
            "previously_purchased": "false",
        },
    },
}

result, err := client.GetBanditAction(
    "shoe-bandit",
    user.ID,
    attributes,
    actions,
    "control",
)
if err != nil {
    log.Printf("Error getting bandit action: %v", err)
    return
}

> if result.Action != "" {

Reviewer: ```suggestion
if result.Action != nil {
```

    showShoeAd(result.Action)
} else {
    showDefaultAd()
}
```

### Subject Context

The subject context contains contextual information about the subject that is independent of bandit actions.
For example, the subject's age or country.

> The subject context has type `Attributes` which has two fields:

Reviewer: ```suggestion
The subject context has type `ContextAttributes` which has two fields:
```

> - `NumericAttributes` (map[string]float64): A map of numeric attributes (such as "age")
> - `CategoricalAttributes` (map[string]string): A map of categorical attributes (such as "country")

Reviewer: ```suggestion
- `Numeric` (map[string]float64): A map of numeric attributes (such as "age")
- `Categorical` (map[string]string): A map of categorical attributes (such as "country")
```
Most examples that reference `NumericAttributes`/`CategoricalAttributes` are off

:::note

> The `CategoricalAttributes` are also used for targeting rules for the feature flag similar to how `subject_attributes` are used with regular feature flags.

Reviewer: You mean `ContextAttributes`?

Author: Yes, that is in fact what I meant.

:::

### Action Contexts

> Next, supply a map with actions and their attributes: `actions: map[string]Attributes`.

Reviewer: ```suggestion
Next, supply a map with actions and their attributes: `actions: map[string]ContextAttributes`.
```

If the user is assigned to the bandit, the bandit selects one of the actions supplied here.
All actions supplied are considered to be valid; if an action should not be shown to a user, do not include it in this map.

The action attributes are similar to the `subject_attributes` but hold action-specific information.

> You can use `Attributes{}` to create an empty attribute context.

Reviewer: ```suggestion
You can use `ContextAttributes{}` to create an empty attribute context.
```

Note that action contexts can contain two kinds of information:
- Action-specific context: e.g., the image aspect ratio of the image corresponding to this action
- User-action interaction context: e.g., there could be a "brand-affinity" model that computes brand affinities of users to brands, and scores of this model can be added to the action context to provide additional context for the bandit.

### Result

The `result` is an instance of `BanditResult`, which has two fields:
- `Variation` (string): The variation that was assigned to the subject

> - `Action` (string): The action that was assigned to the subject (empty string if no action was assigned)

Reviewer: ```suggestion
- `Action` (`*string`): The action that was assigned to the subject (`nil` if no action was assigned)
```

The variation returns the feature flag variation, this can be the bandit itself, or the "status quo" variation if the user is not assigned to the bandit.
If we are unable to generate a variation, for example when the flag is turned off, then the `default` variation is returned.

> In both of those cases, the `Action` is an empty string, and you should use the status-quo algorithm to select an action.

Reviewer: ```suggestion
In both of those cases, the `Action` is nil, and you should use the status-quo algorithm to select an action.
```

When `Action` is not empty, the bandit has selected that action to be shown to the user.

### Status Quo Algorithm

In order to accurately measure the performance of the bandit, we need to compare it to the status quo algorithm using an experiment.
This status quo algorithm could be a complicated algorithm that selects an action according to a different model, or a simple baseline such as selecting a fixed or random action.

When you create an analysis allocation for the bandit and the `Action` in `BanditResult` is empty, implement the desired status quo algorithm based on the `Variation` value.

> ## Debugging
> 
> You may encounter a situation where a bandit assignment produces a value that you did not expect. The SDK provides detailed evaluation information through the `GetBanditActionDetails` method:
> 
> ```go
> evaluation, err := client.GetBanditActionDetails(
>     "shoe-bandit",
>     "test-subject",
>     attributes,
>     actions,
>     "control",
> )
> if err != nil {
>     log.Printf("Error getting bandit details: %v", err)
>     return
> }
> 
> log.Printf("Assignment: %s", evaluation.Variation)
> log.Printf("Action: %s", evaluation.Action)
> log.Printf("Details: %+v", evaluation.EvaluationDetails)
> ```
> 
> For more information on debugging assignments, see [Debugging Flag Assignment](/sdks/sdk-features/debugging-flag-assignment/). 

Reviewer: Not supported
```suggestion
```
